src/python/pants/backend/python/dependency_inference/rules.py

```diff
@@ -289,6 +333,8 @@ def _resolve_single_import(owners, import_name) -> ImportResolveResult:
         if owners.unambiguous:
             return ImportResolveResult(ImportOwnerStatus.unambiguous, owners.unambiguous)
 
+        # this is probably the right place to disambiguate w/ dep rules
+        # after this is too late
```

Comment:
I opted not to put the dep rule filtering logic in `_get_imports_info` to keep the responsibilities of the functions clearer.

Reply:
@lilatomic Hmm. But if I did `_filter_owners_per_import()` inside `_get_imports_info`, then this could report on any possibilities that were discarded due to dep rules.

Is that what you were suggesting in slack?

Reply:
Yes, this was my thought. I think it belongs with the other step that handles disambiguation (based on explicitly provided deps), which would be here. Keeping them together would allow us to distinguish the resolve status between the two modes of disambiguation.
The disambiguation doesn't have to happen here, though. We could move that to somewhere after this function. We could actually use `ImportOwnerStatus.Ambiguous` here, and then filter for that and attempt disambiguation later. For example, inside of `resolve_parsed_dependencies`, which is where you've put the call to `_filter_owners_per_import`. It might make more sense to do it as after the call to `_get_imports_info`.
Am I correct in thinking the "too late" in your comment here refers to the call to `explicitly_provided_deps.maybe_warn_of_ambiguous_dependency_inference`. I think that it might be desirable to separate that warning from `ExplicitlyProvidedDependencies`. In any case, I don't think there's a reason that warning _ought_ to happen here, except that we don't pass that data further down. There's no reason we couldn't propagate this information somehow. A separate disambiguation step would make an obvious place for it, but we could also pass along more information in an `ImportResolveResult` (might be necessary to convert it to a sealed-trait like structure to give the variants extra data slots).

Reply:
> Am I correct in thinking the "too late" in your comment here refers to the call to `explicitly_provided_deps.maybe_warn_of_ambiguous_dependency_inference`.

Yes. That is what I was avoiding.

> I think that it might be desirable to separate that warning from `ExplicitlyProvidedDependencies`. In any case, I don't think there's a reason that warning _ought_ to happen here, except that we don't pass that data further down.

Oh. That makes sense. Since we are adding additional ways to disambiguate, it doesn't really make sense to leave or on `ExplicitlyProvidedDependencies`. Maybe it goes in a static function or a rule_helper. In any case that is an excellent observation that opens up hopefully cleaner ways to implement this feature.

I'll review the rest of your suggestion more carefully and see where that takes us. Thank you.